### PR TITLE
signal-desktop: preload sqlcipher library from nixpkgs

### LINKED ISSUE #124901: signal-desktop cannot open links in the browser
**Describe the bug**
In the latest version (5.3.0) and at least since 5.2.1, I cannot open links in Firefox by clicking them.

This is similar to #78961, however, I have checked that the `nss` packages between Firefox and Signal are the same, using `nix path-info $(which signal-desktop) -r | grep nss-` as described in the wiki article regarding the Discord issue (notably, Discord is also correctly using the same `nss` package and links open properly.

At first I thought it might have been due to using the experimental Wayland support, which also breaks non-text copying and pasting (images, for instance), but running Signal through XWayland results in the same issue.

**To Reproduce**
1. Install `signal-desktop` from the `unstable` channel
2. Try to click a link to open it in the browser
3. The link does not open in the browser

**Expected behavior**
The link should open in the browser

**Screenshots**
N/A

**Additional context**
Running on NixOS 21.11 (unstable), both with the `--enable-features=UseOzonePlatform --ozone-platform=wayland` experimental backend and with the "stable" X backend through XWayland; the result is the same.

**Notify maintainers**
@primeos 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.11.21, NixOS, 21.11pre292520.540dccb2aea (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210503_6d2553a`
 - channels(root): `"nixos-21.11pre292520.540dccb2aea, home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: <nixpkgs/pkgs/applications/networking/instant-messengers/signal-desktop/default.nix>
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by andresilva**
###### Motivation for this change

Fix #124901.

Preloading the packaged node sqlcipher lib was leading to errors like:

 `/bin/sh: symbol lookup error: /nix/store/igb91vpsqwpiw9kpppjhxhhlk4q4b4s5-signal-desktop-5.10.0/lib/Signal/resources/app.asar.unpacked/node_modules/better-sqlite3/build/Release/better_sqlite3.node: undefined symbol: node_module_register` 

which prevented xdg-open calls to work and in turn having links fail to open in the browser. Preloading instead the sqlcipher library from the nixpkgs repo fixes the issue, and I could still open my existing database so I presume encryption is working.

I also needed to add json and fts5 feature support to the sqlcipher library since Signal uses these (these and more are included with the vanilla sqlite package).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix [hunk 1]
```diff
@@ -120,12 +120,8 @@ in stdenv.mkDerivation rec {
     runHook postInstall
   '';
 
-  # Required for $SQLCIPHER_LIB which contains "/build/" inside the path:
-  noAuditTmpdir = true;
-
   preFixup = ''
-    export SQLCIPHER_LIB="$out/lib/Signal/resources/app.asar.unpacked/node_modules/better-sqlite3/build/Release/better_sqlite3.node"
-    test -x "$SQLCIPHER_LIB" # To ensure the location hasn't changed
+    export SQLCIPHER_LIB="${sqlcipher}/lib/libsqlcipher.so"
```
Comment (primeos) [inline pkgs/applications/networking/instant-messengers/signal-desktop/default.nix:124]: We can also drop `$SQLCIPHER_LIB` then (was only a dirty "hack" for the test).

**Review by primeos — CHANGES_REQUESTED** [review]
+1 for the idea but we should think this through first. IIRC we initially decided against using SQLCipher from Nixpkgs due to concerns about compatibility (hopefully this shouldn't be an issue but I haven't looked into it - would be welcome if you have time to look into that and could provide some references) and sticking with upstream's decisions. That said `better_sqlite3.node` is far from ideal...

Also: I'd like to have a dedicated maintainer for `sqlcipher` in Nixpkgs. I can (co-)maintain it but I'd like to have a second maintainer, if possible.

**Comment by andresilva** [comment]
@primeos I looked into it a bit more now. Signal maintains their own fork of `better-sqlite3`, the only substantial change is this commit https://github.com/signalapp/better-sqlite3/commit/5acdaf9d035388b6736f50d96dc8b034219a367e where they replace the sqlite source code with the sqlcipher source. I've downloaded the sqlcipher source tarball they have there:

```❯ rg "CIPHER_VERSION_NUMBER"
21451:#define CIPHER_VERSION_NUMBER 4.4.3
```

This is the same version as the sqlcipher package in nixpkgs. Apart from that their build has more features enabled than we're currently using for the sqlcipher package. 

Their:

```
    'SQLITE_LIKE_DOESNT_MATCH_BLOBS',
    'SQLITE_THREADSAFE=2',
    'SQLITE_USE_URI=0',
    'SQLITE_DEFAULT_MEMSTATUS=0',
    'SQLITE_OMIT_DEPRECATED',
    'SQLITE_OMIT_GET_TABLE',
    'SQLITE_OMIT_TCL_VARIABLE',
    'SQLITE_OMIT_PROGRESS_CALLBACK',
    'SQLITE_OMIT_SHARED_CACHE',
    'SQLITE_TRACE_SIZE_LIMIT=32',
    'SQLITE_DEFAULT_CACHE_SIZE=-16000',
    'SQLITE_DEFAULT_FOREIGN_KEYS=1',
    'SQLITE_DEFAULT_WAL_SYNCHRONOUS=1',
    'SQLITE_ENABLE_COLUMN_METADATA',
    'SQLITE_ENABLE_UPDATE_DELETE_LIMIT',
    'SQLITE_ENABLE_STAT4',
    'SQLITE_ENABLE_FTS5',
    'SQLITE_ENABLE_JSON1',
    'SQLITE_ENABLE_RTREE',
    'SQLITE_INTROSPECTION_PRAGMAS',

    # SQLCipher-specific options
    'SQLITE_HAS_CODEC',
    'SQLITE_TEMP_STORE=2',
    'SQLITE_SECURE_DELETE',
```

Ours:
```
    "-DSQLITE_ENABLE_COLUMN_METADATA=1"
    "-DSQLITE_ENABLE_FTS5"
    "-DSQLITE_ENABLE_JSON1"
    "-DSQLITE_ENABLE_UNLOCK_NOTIFY=1"
    "-DSQLITE_HAS_CODEC"
    "-DSQLITE_SECURE_DELETE=1"
```

I have been using Signal for the past hours and everything seems to be working fine so far.

> Also: I'd like to have a dedicated maintainer for `sqlcipher` in Nixpkgs. I can (co-)maintain it but I'd like to have a second maintainer, if possible.

I have never used `sqlcipher` and only looked into it now to fix this Signal issue. I am OK with being added as co-maintainer but just wanted to let you know I barely know anything about these packages :P

File: pkgs/applications/networking/instant-messengers/signal-desktop/default.nix (hunk 1 above)
Comment (andresilva) [reply pkgs/applications/networking/instant-messengers/signal-desktop/default.nix]: Done.

**Comment by andresilva** [comment]
@primeos I opted to not change the sqlcipher package directly and instead override it for signal to make sure we can use the exact same build features as the upstream signal npm package. I think this shouldn't cause any trouble now, we're building the same source with the same features enabled, we just need to keep an eye on upstream for changes going forward unfortunately.

**Comment by andresilva** [comment]
@primeos Any thoughts on this? I'm using this since I opened the PR, and I've been updating it locally with the new Signal versions. No issues so far.

**Review by primeos — APPROVED** [review]

**Pull request merged**